Tidyups

purcell opened the pull request:
Hi Sibi, I was trying out `justl` but found it didn't work for me because I use [`envrc.el`](https://github.com/purcell/envrc) with per-project tooling and my `just` isn't normally installed globally. Ultimately I expect I'd need to use [`inheritenv`](https://github.com/purcell/inheritenv) to make sure the `justl.el` commands pick up the right local environment. I started digging into the code and found myself tweaking a few things, but still haven't fixed the root problem. In the meantime, here are a bunch of little tweaks and improvements, in separate commits so you follow the reasoning. (I've tested this briefly by installing `just` globally.)

psibi commented:
Thanks @purcell! I still haven't given through it, but can you check why the CI is failing ?

purcell commented:
Yup, done, tests work locally for me now.

psibi commented:
Thanks for the PR! This looks great.

> I started digging into the code and found myself tweaking a few things, but still haven't fixed the root problem.

Have you fixed the root problem ? If no, do you want to it as part of a new PR or use the same one ?  (The only thing that's remaining from my end is that I need to do the testing of TRAMP integration manually)

purcell commented:
I'd be happy to submit a subsequent PR. If you're happy with this one, go ahead and merge. I have a couple more refactorings in mind, but might not get to them soon. And I've worked around the root problem for now by installing globally, though I still hope to resolve it in due course.

purcell commented:
I fixed the root issue, and also switched things to use `justl --dump --dump-format=json` to make parsing info about recipes easier.

purcell commented:
There's info in the json about the parameters too, so more code around `justl-exec-recipe` could be tidied up in this way too. Done for today, though.

purcell commented:
(Tests broken again with latest changes, will resolve and confirm when done.)

purcell commented:
Okay, I replaced all of the custom parsing with `just`'s own parsing, and now info about the recipes gets kept in the list view for convenient use when people initiate actions. Can't guarantee I haven't messed anything up, but I think this is in pretty good shape, and having updated the tests appropriately, they pass.

purcell commented:
(And happy to be pinged if any issues arise.)

psibi commented:
@purcell Sorry, it took me a while to get back to this. From my testing, the tramp integration seems to be broken. When I invoke `justl`, I get this error message:

```
helm-M-x-execute-command: Couldn’t read /docker:pedantic_wozniak:/home/sibi/justfile: just exited with code 1
```

This is how I test the TRAMP integration: https://github.com/psibi/justl.el/blob/master/test/README.org (I have the corresponding Dockerfile in the test directory). Let me know if you have difficulty reproducing this and I can expand the documentation.

purcell commented:
That should do it. I had to fix up the dockerfile because it didn't work for me, but otherwise your instructions were great.

purcell commented:
Ah no, not quite yet. Will look again in a bit.

purcell commented:
There we go, I had broken a test along the way, but everything's in good shape now, I think.

psibi commented:
I tested it more and I believe some more fixes are required:

- In the just buffer, the recipes comes in an unordered fashion. The earlier behavior was to match the same order as in the actual justfile. Probably you can use something like this to figure out the order: `just --summary --unsorted`
- When I try to execute a recipe which has arguments, the completion given by it is not correct. Right now it tries to fill with some elisp code. But the behavior should be this: If a default argument is specified in justfile, that should be populated. If not, nothing should be populated. (You can also check the behavior from the current master branch)
- When I try to execute `M-x justl-exec-recipe-in-dir`, it fails with this error message:

```
helm-M-x-execute-command: Wrong type argument: listp
```

purcell commented:
> tested it more and I believe some more fixes are required:
> 
>     * In the just buffer, the recipes comes in an unordered fashion. The earlier behavior was to match the same order as in the actual justfile. Probably you can use something like this to figure out the order: `just --summary --unsorted`

Retaining the original order is not supported by the `--dump` command, but the resulting list _is_ sorted alphabetically (both by the elisp _and_ – apparently – by `--dump`), which I think is preferable anyway. Note that the interface of `just` itself defaults to alphabetical, for both `just --list` and `just --summary`. The `--unsorted` flag has no effect on `--dump`.

>     * When I try to execute a recipe which has arguments, the completion given by it is not correct. Right now it tries to fill with some elisp code. But the behavior should be this: If a default argument is specified in justfile, that should be populated. If not, nothing should be populated. (You can also check the behavior from the current master branch)

Fixed now, but I'd note that the existing code probably isn't ideal anyway. Consider:
```
baz := "hello"
test foo bar=baz:
   ...
```
Here you can't really insert the default value such that the user can just hit return. What you'd probably really want would be to show the default in the prompt, e.g. `Just arg for `foo' [default: blah]: ` and then hit enter to select the default, but pass *nothing* through to the `just` invocation, ie. actually pass fewer arguments.


>     * When I try to execute `M-x justl-exec-recipe-in-dir`, it fails with this error message:

Fixed.

psibi commented:
> Retaining the original order is not supported by the --dump command, but the resulting list is sorted alphabetically (both by the elisp and – apparently – by --dump)

Yeah, I know. That's why I suggested the `--summary --unsorted` option to get the recipe names separately and preserve ordering.

> which I think is preferable anyway.

Unfortunately, my use case requires it. To give you an explicit example, I have this justfile which I use often use during testing: https://github.com/psibi/app_k8s/blob/master/justfile
The different targets there has same pattern: build, apply and remove. If they don't come in the same order as specified in the justfile, I would have to keep hunting from the huge number of similar recipes present.

> The --unsorted flag has no effect on --dump.

Yeah, that's unfortunate. You have to do another call to `just --summary --unsorted` to figure out the proper order in the current setup.

> Fixed now, but I'd note that the existing code probably isn't ideal anyway.

Yeah, agree. With your changes, I think implementing something like that would be easier in future since we can resolve variables using the json dump that we are getting.

> Fixed

Thank you! My elisp has become better as I have learnt good amount of things in this PR.

purcell commented:
almost there, just working around what appears to be an emacs bug

purcell commented:
Alrighty, that's done. 😅

psibi commented:
@purcell Just tested and this looks great. I believe there is just one more change: When I try to execute a recipe which has an argument, it prompts like this:

```
Just arg for `version':
```

Can it be instead like this:

```
Just arg for 'version':
```

The only difference is that in the second version there is single quote. Also, can you update the changelog too (Probably add a unreleased section) ?

purcell commented:
That's a pretty standard style for printing these things in Emacs, so I typed that backquote deliberately, but equally happy to change it. Probably be tomorrow. :D

purcell commented:
Done now

psibi commented:
Thank you!

purcell commented:
🤝
